src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -487,54 +430,52 @@ private MethodBuilder AddMethodImpl(MethodInfo mi)
-                GenericArray<object> packedArr = new GenericArray<object>(il, PackedArgs.PackedTypes.Length);
-
-                // packed[PackedArgs.DispatchProxyPosition] = this;
-                packedArr.BeginSet(PackedArgs.DispatchProxyPosition);
+                // MethodInfo methodInfo = _methodInfos[methodInfoIndex];
+                LocalBuilder methodInfoLocal = il.DeclareLocal(typeof(MethodInfo));
                 il.Emit(OpCodes.Ldarg_0);
-                packedArr.EndSet(typeof(DispatchProxy));
-
-                // packed[PackedArgs.DeclaringTypePosition] = typeof(iface);
-                MethodInfo Type_GetTypeFromHandle = typeof(Type).GetRuntimeMethod("GetTypeFromHandle", new Type[] { typeof(RuntimeTypeHandle) })!;
-                _assembly.GetTokenForMethod(mi, out Type declaringType, out int methodToken);
-                packedArr.BeginSet(PackedArgs.DeclaringTypePosition);
-                il.Emit(OpCodes.Ldtoken, declaringType);
-                il.Emit(OpCodes.Call, Type_GetTypeFromHandle);
-                packedArr.EndSet(typeof(object));
-
-                // packed[PackedArgs.MethodTokenPosition] = iface method token;
-                packedArr.BeginSet(PackedArgs.MethodTokenPosition);
-                il.Emit(OpCodes.Ldc_I4, methodToken);
-                packedArr.EndSet(typeof(int));
-
-                // packed[PackedArgs.ArgsPosition] = args;
-                packedArr.BeginSet(PackedArgs.ArgsPosition);
-                argsArr.Load();
-                packedArr.EndSet(typeof(object[]));
+                il.Emit(OpCodes.Ldfld, _fields[MethodInfosFieldAndCtorParameterIndex]); // MethodInfo[] _methodInfos
+                il.Emit(OpCodes.Ldc_I4, methodInfoIndex);
+                il.Emit(OpCodes.Ldelem_Ref);
+                il.Emit(OpCodes.Stloc, methodInfoLocal);
 
-                // packed[PackedArgs.GenericTypesPosition] = mi.GetGenericArguments();
                 if (mi.ContainsGenericParameters)
                 {
-                    packedArr.BeginSet(PackedArgs.GenericTypesPosition);
+                    // methodInfo = methodInfo.MakeGenericMethod(mi.GetGenericArguments());
+                    il.Emit(OpCodes.Ldloc, methodInfoLocal);
+
                     Type[] genericTypes = mi.GetGenericArguments();
                     GenericArray<Type> typeArr = new GenericArray<Type>(il, genericTypes.Length);
                     for (int i = 0; i < genericTypes.Length; ++i)
                     {
                         typeArr.BeginSet(i);
                         il.Emit(OpCodes.Ldtoken, genericTypes[i]);
-                        il.Emit(OpCodes.Call, Type_GetTypeFromHandle);
+                        il.Emit(OpCodes.Call, s_getTypeFromHandleMethod);
                         typeArr.EndSet(typeof(Type));
                     }
                     typeArr.Load();
-                    packedArr.EndSet(typeof(Type[]));
+
+                    il.Emit(OpCodes.Callvirt, s_makeGenericMethodMethod);
```

Comment:
`MakeGenericMethodMethod` is slow API. What does the performance of this path looks like with this change?

Reply:
The performance of this code path shouldn't be changed because the call to MakeGenericMethod is just moved from inside our code to the generated code.

Here's where it was called before, during each invocation:

https://github.com/dotnet/runtime/blob/5078dd61a81c88dbdd131a77b8148b4dee3803fa/src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs#L139-L147

I added a generic method benchmark to the above benchmark results.